### ISSUE TITLE
capnp::writeMessageToFd inconsistent behavior on MacOS & Windows with large messages

### ISSUE THREAD

**Issue by vlovich** [capnproto/capnproto#1128]
TLDR: I think it's probably a good idea to restrict the maximum write size issued on MacOS to 2GB & on Windows to 4GB. Large writes seem to be broken on these platforms (throws an exception on MacOS & I *think* should result in silent data corruption on Windows). I think it's also entirely possible that even the fact that MacOS throws an exception vs memory corruption is similarly an artifact of the size I chose (i.e. I think if you end up writing some amount you should be able to get the signed sum to overflow back above 0 & libc might not notice the overflow?).

# MacOS
From the [manual page](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/write.2.html):

```
     [EINVAL]           The sum of the iov_len values in the iov array over-flows overflows
                        flows a 32-bit integer.
```

This appears to apply to `write` to my testing as well. To reproduce, create a message with a `List(Data)` field that has 10 512 MB elements (~5 GB) & call `writeMessageToFd`.

# 64-bit Linux:
This doesn't appear to be an issue on 64-bit Linux which [defines](https://linux.die.net/man/2/writev) overflow in terms of `ssize_t`:

```
EINVAL
The sum of the iov_len values overflows an ssize_t value. Or, the vector count iovcnt is less than zero or greater than the permitted maximum.
```

# 32-bit Linux

Probably the same bug as MacOS since ssize_t will be 32-bit signed. Unclear how important this platform is but probably fixing MacOS would fix this too.

# Windows
As best I can tell there's a similar [limitation on Windows](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/write?view=msvc-160) around the max number of bytes being representable as unsigned 32-bit (~4gb). The issue here I think (not tested) is that cap'n'proto has an implicit [size_t -> unsigned int overflow](https://github.com/capnproto/capnproto/blob/1f892a3ca38446426c816dcf9dfd3d1b31053a09/c%2B%2B/src/kj/miniposix.h#L60) as best I can tell which should result in corruption (again - untested but seems conceptually right since I saw large single contiguous writes even with the default writev OutputStream implementation).

**Comment by kentonv** [comment]
This should be easy to fix by truncating the `iov` list at 2^31-1 bytes. The code already has to handle `writev()` returning a partial write, in which case it has to retry the rest of the write. In fact, I'm surprised MacOS doesn't just do this truncation itself and return MAX_INT as the number of bytes written.

Regarding other platforms, I don't think there's a bug:
* 32-bit Linux: It would be impossible to overflow here because the maximum size of the buffer being written is inherently limited by the process's virtual address space.
* Windows: We don't do a gather-write there (Windows doesn't even have a usable gather-write API), hence each `write()` would be writing a single segment. Cap'n Proto segments are limited to 2^29-1 words in size, or 2^32-8 bytes. So I think, at least in Cap'n Proto's usage, we can't end up overflowing there, although I agree other uses of the KJ APIs could in theory overflow and that seems like a bug worth fixing.

**Comment by vlovich** [comment]
You're obviously right about this not being an issue for 32-bit operating systems. I assumed this was an issue for Windows because I saw problems removing the writev on MacOS but it sounds reasonable that it's only an issue because writes on MacOS are capped at ~2GB & while Windows' 4GB cap on writes is outside the maximum contiguous write Cap'n'Proto would issue. Let me see if I can throw together a fix, although I don't have any access to non-Mac platforms at the moment so my ability to test things out on my own is limited. I'm thinking the fix for Windows is as simple as saturating the write at 4GB since partial writes are presumably the writers responsibility, right?.

**Comment by kentonv** [comment]
> I'm thinking the fix for Windows is as simple as saturating the write at 4GB since partial writes are presumably the writers responsibility, right?

Right -- same as how I'd suggest fixing `writev()` on mac.

**Comment by vlovich** [comment]
Hmm.... the fix for write seems straightforward to me but the writev fix seems less so, but it's likely I'm missing something obvious. My reading is that `FdOutputStream` calls  ::writev directly so there's nowhere to interpose (trivially solved). The complexity is I'd have to modify the iovec before passing it to writev & carefully restore it afterward (carefully accounting that even a single entire iov may exceed 2GB). Am I missing something?

My rough solution is to solve it through a miniposix.h writev interposition. Something like:

```
ssize_t miniposix::writev(int fd, iovec* iov, int niovec) {
#if __APPLE__
    // Workaround for writev POSIX bug on Apple (filed as FB8934446) that prevents writes > 2GB.
    ssize_t total = 0;
    int truncatedEntry;
    size_t truncatedEntryLen;
    for (truncatedEntry = 0; truncatedEntry < niovec; ++truncatedEntry) {
      total += iov[truncatedEntry].iov_len;
      if (total > INT_MAX) {
         truncatedEntryLen = iov[truncatedEntry].iov_len;
         iov[truncatedEntry].iov_len = INT_MAX - total;
         break;
      }
    }
    ssize_t result = ::writev(fd, iov, truncatedEntryLen);
    if (truncatedEntry != niovec) {
      iov[truncatedEntry].iov_len = truncatedEntryLen;
    }
    return result;
#else
    return ::writev(fd, iov, niovec);
#endif
}
```

Is this roughly what you were thinking of or did you have a simpler solution in mind?

**Comment by kentonv** [comment]
Fixed by #2198.
